docs/src/tutorials/getting_started/design_patterns_for_larger_models.jl

```diff
@@ -0,0 +1,596 @@
+#  * it is easy to introduce errors, e.g., having `profit` and `weight` be
+#    vectors of different lengths, or not match `N`
+#  * the solution, `x[i]`, is hard to interpret without knowing the order in
+#    which we provided the data.
+
+# ## Wrap the model in a fuction
+
+# A good next step is to wrap your model in a function. This is useful for a few
+# reasons:
+#  * it removes global variables
+#  * it encapsulates the JuMP model and forces you to clarify your inputs and
+#    outputs
+#  * we can add some error checking.
+
+function solve_knapsack_1(profit::Vector, weight::Vector, capacity::Real)
+    @assert length(profit) == length(weight)
+    N = length(weight)
+    model = Model(GLPK.Optimizer)
+    @variable(model, x[1:N], Bin)
+    @objective(model, Max, sum(profit[i] * x[i] for i in 1:N))
+    @constraint(model, sum(weight[i] * x[i] for i in 1:N) <= capacity)
+    optimize!(model)
+    return value.(x)
+end
+
+solve_knapsack_1([5, 3, 2, 7, 4], [2, 8, 4, 2, 5], 10)
+
+# ## Create better data structures
+
+# Although we can check for errors like mis-matched vector lengths, if you start
+# to develop models with a lot of data, keeping track of vectors and lengths and
+# indices is fragile and a common source of bugs. A good solution is to use
+# Julia's type system to create an abstraction over your data.
+
+# For example, we can create a struct that represents a single object:
+
+struct KnapsackObject
+    profit::Float64
+    weight::Float64
+end
+
+# as well as a struct that holds a dictionary of objects and the knapsack's
+# capacity:
+
+struct KnapsackData
+    objects::Dict{String,KnapsackObject}
+    capacity::Float64
+end
+
+# Here's what our data might look like now:
+
+objects = Dict(
+    "apple" => KnapsackObject(5.0, 2.0),
+    "banana" => KnapsackObject(3.0, 8.0),
+    "cherry" => KnapsackObject(2.0, 4.0),
+    "date" => KnapsackObject(7.0, 2.0),
+    "eggplant" => KnapsackObject(4.0, 5.0),
+)
+data = KnapsackData(objects, 10.0)
+
+# If you want, you can add custom printing to make it easier to visualize:
+
+function Base.show(io::IO, data::KnapsackData)
+    println(io, "A knapsack with capacity $(data.capacity) and possible items:")
+    for (k, v) in data.objects
+        println(
+            io,
+            "  $(rpad(k, 8)) : profit = $(v.profit), weight = $(v.weight)",
+        )
+    end
+    return
+end
+
+data
+
+# Then, we can re-write our `solve_knapsack` function to take our `KnapsackData`
+# as input:
+
+function solve_knapsack_2(data::KnapsackData)
+    model = Model(GLPK.Optimizer)
+    @variable(model, x[keys(data.objects)], Bin)
+    @objective(model, Max, sum(v.profit * x[k] for (k, v) in data.objects))
+    @constraint(
+        model,
+        sum(v.weight * x[k] for (k, v) in data.objects) <= data.capacity,
+    )
+    optimize!(model)
+    return value.(x)
+end
+
+solve_knapsack_2(data)
+
```

Comment:
Could add something like:

This approach also allows us to write functions that do error checking
on the data:
```
function check_knapsack_data(data::KnapsackData)
    for (name,object) in data.objects
         @assert object.profit >= 0
         @assert object.weight >  0
    end
end
```

Reply:
I added a constructor for `KnapsackObject` with an error check